### PR TITLE
enhance performance by ignoring rules that are never used #12592

### PR DESCRIPTION
**Pull request by dynasource**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | not likely |
| Tests pass? | yes |
| Fixed issues | #12592 |

### CODE REVIEW

File: framework/filters/AccessControl.php [hunk 1]
```diff
@@ -111,10 +96,13 @@ public function init()
      */
     public function beforeAction($action)
     {
-        $user = $this->user;
+        $user = $this->user = Instance::ensure($this->user, User::className());
         $request = Yii::$app->getRequest();
         /* @var $rule AccessRule */
         foreach ($this->rules as $rule) {
+            if (!is_object($rule)) {
+                $rule = Yii::createObject(array_merge($this->ruleConfig, $rule));
```
Comment (SilverFire) [inline framework/filters/AccessControl.php]: `$rule` object must be set to `$this->rules` property to be reused in future

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Comment by SilverFire** [comment]
Could you add a CHANGELOG line?

**Review by SilverFire — APPROVED** [review]
Add CHANGELOG and fine to go :)

**Review by SilverFire — APPROVED** [review]

**Comment by dynasource** [comment]
I repackaged it into 1 commit. Thx!

**Review by SilverFire — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by cebe** [comment]
As this could break BC for people overriding init() method and it is a critical part for website security I'd prefer to move this PR to 2.1.

**Comment by dynasource** [comment]
agreed

**Comment by cebe** [comment]
@dynasource can you make the same changes based on the 2.1 branch and send the PR to that branch?

**Review by cebe — APPROVED** [review]

**Comment by SilverFire** [comment]
@dynasource Could you update UPGRADE.md and mention this change as a potential BC break?

**Comment by SilverFire** [comment]
:+1:

**Pull request merged**